### ISSUE TITLE
Scrollbar jumps up after selecting an item in a Tree.

### ISSUE THREAD

**Issue by XeptionE** [gwtproject/gwt#9475]
GWT 2.7, newest Chrome-Version (55.0)

I narrowed this to the code, which is responsible to exclude the focus() (and blur()) functions from the event flow for safari browsers. While debugging all is fine, but then somehow GWT wants to change the focus and the scrollbar jumps to the top.

Bug was introduced some weeks ago with a new chrome browser (I think before ~52.0 all was fine). I didn't change anything in my code. Firefox works well.

My current hack is to set "display: block !important;" on all ".gwt-TreeItem". I don't know why, but this helps. However it is not perfect, because sometimes the lists still jumps (to a random item(!!!!)).

But my hack prevents <input>-Element (which is part of the tree item content) to be focussed programmatically, after I have added a new item to the list. Without my hack programmatical focus works. I don't know, why switching from display-inline to block of a treeitem make that big change in logical behaviour (and even the layout stays unchanged with this ^^).

Because I didn't experience any problems while surfing in web, I think that it is a GWT bug.

**Comment by tbroyer** [comment]
AFAIK this has always been the case. See #4456 (possibly others too)

**Comment by XeptionE** [comment]
No, it was not the case for the last 2 years. :) Only since some weeks (maybe a few months?) I see the problem.

**Comment by jnehlmeier** [comment]
Hm can't really produce any jumping using http://samples.gwtproject.org/samples/Showcase/Showcase.html#!CwTree

Tried it with Safari 10 and Chrome 57 and opened a few of the first items, then scrolled down and selected/open/closed one of the bottom items.

Maybe it is some code in your app that is causing the jumping?

**Comment by ibaca** [comment]
I recently see a similar problem with a hover style with an :after
{content:""}, this produces the scroll reset in chrome.

**Comment by XeptionE** [comment]
You can reproduce it using my projects homepage. Follow this link: http://buildhome.eu/#installation

You will see a list on the left. Scroll down and select an element -> it will NOT(!) scroll to the top.


Now open in Chrome the debugging tools (F12). Hit Ctrl+Shift+C (element inspector) and select an item in the scrolled-list. In the element-Tree you will see a "<div class="gwt-TreeItem" ...>" as a parent element of your inspected element. Select this element and on the right side in the development tools you will see the CSS styles set on GWT-TreeItems (it should be the second selector ".gwt-TreeItem { ... }"). There you will find my "hack" = display: block !important;

Now disable the hack (by clicking on the checkbox near the "display: block !important" CSS rule). From now on the list will jump to top, if you select an item. THIS WAS NOT THE CASE WITH AN OLDER CHROME (maybe <= 52.0, don't know exact version, but it must be a newer change). However, my Hack is not perfect. If the tree items contain an <input> element and I add an item to the tree programmaticaly, then using Scheduler.get().runDeffered... it is not possible to focus the <input>-element. Without the hack it is possible.

**Comment by XeptionE** [comment]
I have found the problem, maybe ^^:

- FocusImplSafari.java (its a GWT toolchain file)
- in the method "focus(Element elem) on the line "elem.focus()".

Before this line all is fine with my application. But I don't know how to prevent GWT from executing this focus code, beacuse it happes AFTER all application calls already done. The focussed element is the first <div>element of the tree-<div>, it is generated by GWT toolkit in the Tree.java code in its init()-method (I think, this is the "focusable" element), it has the following content in the DOM-Tree:

<div tabindex="0" hidefocus="true" role="tree" style="font-size: 0px; position: absolute; outline: 0px; z-index: -1; left: 0px; top: 0px;"><input type="text" tabindex="-1" role="presentation" style="opacity: 0; height: 1px; width: 1px; z-index: -1; overflow: hidden; position: absolute;"></div>

After I removed this item, the scrolling to top does not happen anymore (don't know, what it is for ^^, in code I see some comments about IE6/7 issues).


Now I don't have a hack for you, but I imagine, that programmatically removing the first child of the tree element after its creating would help (check for role="tree" to be sure). It would be nice, if the GWT team could look into this :) because I love GWT, which I preffer over google dart (which is nice too), and ofcourse direct javascript hacking.

**Comment by jnehlmeier** [comment]
The root cause is that the focusable element (the first child of the Tree) is updated with a wrong absolute positioning after you have selected an element (after removing your hack). Normally this focusable element should have the absolute position matching the position of the clicked element.

If the tree item has `display:inline` the focusable element gets a position with `top:0; left:0;` and thus the scroll panel scrolls back to top. With your hack the `top` and `left` positions are correctly applied and no scrolling occurs. You can see this in the element inspector, just look at the first `div` child of the tree which has the tabindex and style attribute.

Basically the method `moveFocus()` does something wrong as its responsible to move the focusable element at the position of the selected tree item: 

https://github.com/gwtproject/gwt/blob/master/user/src/com/google/gwt/user/client/ui/Tree.java#L1157

**Comment by XeptionE** [comment]
Yes, I have discovered it some minutes before your comment too :) Thank you. Look at my comment, there is a code, which prevents the moveFocus.

Maybe it has something to do with this: https://github.com/jquery/jquery/issues/2889

**Comment by leforthomas** [comment]
I have the same issue, was fine until a month or two ago and now it jumps back to the beginning of the tree. As I can have very large trees it's really not acceptable for the users.
I've read the various comments (thanks!) but I don't know if there is a quick fix that can be applied?

**Comment by bricebby** [comment]
Same problem here. I was able to workaround this by overriding the setFocus method to don't call the native focus method. So far seems it's breaking only the navigation in the tree with the keyboard.
```
tree = new Tree() {  
            @Override
            public void setFocus(boolean focus) {
                // Don't do anything
            }
        };
```

**Comment by leforthomas** [comment]
Yep, that does the trick, thanks!

**Comment by XeptionE** [comment]
You also can try this (I have mentioned this earlier in my posts as workaround):

.gwt-TreeItem {
	display: block !important;
}

(yes, you need to insert this to one of your global CSS ^^)

**Comment by jnehlmeier** [comment]
Chrome Bug for reference: https://bugs.chromium.org/p/chromium/issues/detail?id=681382

**Comment by tbroyer** [comment]
So, should we close a Browser-Quirk? or are there known workarounds? (using `getBoundingClientRect`?)

**Comment by Polybios90** [comment]
If you have set Tree.scrollOnSelectEnabled to false, then this issue also occurs. I.e., because the focusable item will never be updated on selection change and via setFocus the focus will jump up to the original top=0 of the focusable item.
(https://github.com/gwtproject/gwt/blob/master/user/src/com/google/gwt/user/client/ui/Tree.java#L1165). Besides I use 
```
.gwt-TreeItem {
	display: inline-table !important;
}
```
because display: block !important caused a bug concerning the background of the selected item, when I use:
```
.gwt-TreeItem-selected {
	background: #ebeff9;
}
```
as in the show case [gwt tree showcase](http://samples.gwtproject.org/samples/Showcase/Showcase.html#!CwTree)
